File: Examples/WebAssembly.Examples/CanvasInterop.cs
using System.Runtime.InteropServices.JavaScript;

public partial class CanvasInterop
{
    [JSImport("init", "canvas-interop")]
    public static partial void Init();

    [JSImport("getWidth", "canvas-interop")]
    public static partial int GetWidth();

    [JSImport("getHeight", "canvas-interop")]
    public static partial int GetHeight();

    [JSImport("clear", "canvas-interop")]
    public static partial void Clear(byte r, byte g, byte b, byte a);

    [JSImport("drawBorder", "canvas-interop")]
    public static partial void DrawBorder(byte r1, byte g1, byte b1, byte a1, byte r2, byte g2, byte b2, byte a2, float x, float y, float w, float h, int border_size);

    [JSImport("drawRectangle", "canvas-interop")]
    public static partial void DrawRectangle(byte r, byte g, byte b, byte a, float x, float y, float w, float h);

    [JSImport("pushClip", "canvas-interop")]
    public static partial void PushClip(float x, float y, float w, float h);

    [JSImport("popClip", "canvas-interop")]
    public static partial void PopClip();

    [JSImport("createCanvas", "canvas-interop")]
    public static partial int CreateCanvas(int width, int height);

    [JSImport("destroyCanvas", "canvas-interop")]
    public static partial void DestroyCanvas(int id);

    [JSImport("setCanvasPixels", "canvas-interop")]
    public static partial void SetCanvasPixels(int id, int width, int height, [JSMarshalAs<JSType.MemoryView>] Span<byte> pixels);

    [JSImport("copyCanvasPixels", "canvas-interop")]
    public static partial void CopyCanvasPixels(int id, int fromX, int fromY, int fromWidth, int fromHeight, int toX, int toY, int toWidth, int toHeight, byte r, byte g, byte b, byte a);

    [JSImport("getEventsCount", "canvas-interop")]
    public static partial int GetEventsCount();

    [JSImport("getEvent", "canvas-interop")]
    public static partial int GetEvent(int index);

    [JSImport("getEventProperty", "canvas-interop")]
    public static partial int GetEventProperty(int index, string property);

    [JSImport("getEventPropertyString", "canvas-interop")]
    public static partial int GetEventPropertyString(int index, string property, [JSMarshalAs<JSType.MemoryView>] Span<int> buffer); //Returns string length

    [JSImport("clearEvents", "canvas-interop")]
    public static partial void ClearEvents();

    [JSImport("setCursor", "canvas-interop")]
    public static partial void SetCursor(string cursor);
}

File: Examples/WebAssembly.Examples/canvas-interop.js

let canvas;
let ctx;

const half_pixel = 0.5;

export function init() {
    canvas = document.getElementById("myCanvas");
    console.log(canvas);
    ctx = canvas.getContext("2d");
    ctx.imageSmoothingEnabled = false;
}

export function getWidth() {
    return canvas.width;
}

export function getHeight() {
    return canvas.height;
}

export function clear(r, g, b, a) {
    ctx.fillStyle = `rgba(${r}, ${g}, ${b}, ${a / 255})`;
    ctx.fillRect(0, 0, canvas.width, canvas.height);
}

export function drawBorder(r1, g1, b1, a1, r2, g2, b2, a2, x, y, w, h, border_size) {
    
    if (a1 != 0) {
        ctx.fillStyle = `rgba(${r1}, ${g1}, ${b1}, ${a1 / 255})`;
        ctx.fillRect(x, y, w, h);
    }

    if (a2 != 0) {
        ctx.strokeStyle = `rgba(${r2}, ${g2}, ${b2}, ${a2 / 255})`;
        ctx.lineWidth = border_size;
        ctx.strokeRect(x + border_size / 2, y + border_size / 2, w - 1, h - 1);
    }
}

export function drawRectangle(r1, g1, b1, a1, x, y, w, h) {
    if (a1 != 0) {
        ctx.fillStyle = `rgba(${r1}, ${g1}, ${b1}, ${a1 / 255})`;
        ctx.fillRect(x, y, w, h);
    }
}

export function pushClip(x, y, w, h) {
    ctx.save();
    ctx.rect(x, y, w, h);
    ctx.clip();
}

export function popClip() {
    ctx.restore();
}